app/services/ItemsCachingService.scala

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services
+import com.google.inject.Inject
+import config.AppConfig
+import controllers.util.CacheIdGenerator.itemsId
+import forms.supplementary.ItemType.IdentificationTypeCodes._
+import forms.supplementary.{CommodityMeasure, ItemType, PackageInformation}
+import javax.inject.Singleton
+import models.DeclarationFormats._
+import models.declaration.supplementary.{AdditionalInformationData, DocumentsProducedData, ProcedureCodesData}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.http.cache.client.CacheMap
+import uk.gov.hmrc.wco.dec._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class ItemsCachingService @Inject()(cacheService: CustomsCacheService)(appConfig: AppConfig) {
+
+  def generatePackages(cachedData: CacheMap) =
```

Comment:
There are some very 'suspicious' return types in these methods (like `Option[Seq[_]]`).
I think this is because of nested `map` methods, while there should be `flatMap`.

Reply:
We need to handle no data scenario and I have this so we can update these methods with a common error scenario